polygon/heimdall/simulator/heimdall_simulator.go

```diff
@@ -0,0 +1,169 @@
+package simulator
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/ledgerwatch/log/v3"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/ledgerwatch/erigon-lib/chain/snapcfg"
+	"github.com/ledgerwatch/erigon-lib/common/background"
+	"github.com/ledgerwatch/erigon-lib/downloader/snaptype"
+	"github.com/ledgerwatch/erigon/eth/ethconfig"
+	"github.com/ledgerwatch/erigon/polygon/heimdall"
+	"github.com/ledgerwatch/erigon/turbo/snapshotsync/freezeblocks"
+)
+
+type HeimdallSimulator struct {
+	ctx                context.Context
+	knownSnapshots     *freezeblocks.RoSnapshots
+	activeSnapshots    *freezeblocks.RoSnapshots
+	knownBorSnapshots  *freezeblocks.BorRoSnapshots
+	activeBorSnapshots *freezeblocks.BorRoSnapshots
+	blockReader        *freezeblocks.BlockReader
+	logger             log.Logger
+	downloader         *TorrentClient
+
+	nextSpan                  uint64
+	lastDownloadedBlockNumber uint64
+}
+
+type IndexFnType func(context.Context, snaptype.FileInfo, string, *background.Progress, log.Lvl, log.Logger) error
+
+func NewHeimdall(ctx context.Context, chain string, snapshotLocation string, logger log.Logger) (HeimdallSimulator, error) {
+	cfg := snapcfg.KnownCfg(chain)
+
+	knownSnapshots := freezeblocks.NewRoSnapshots(ethconfig.Defaults.Snapshot, snapshotLocation, 0, logger)
+	knownBorSnapshots := freezeblocks.NewBorRoSnapshots(ethconfig.Defaults.Snapshot, snapshotLocation, 0, logger)
+
+	files := make([]string, 0, len(cfg.Preverified))
+
+	for _, item := range cfg.Preverified {
+		files = append(files, item.Name)
+	}
+
+	knownSnapshots.InitSegments(files)
+	knownBorSnapshots.InitSegments(files)
+
+	activeSnapshots := freezeblocks.NewRoSnapshots(ethconfig.Defaults.Snapshot, snapshotLocation, 0, logger)
+	activeBorSnapshots := freezeblocks.NewBorRoSnapshots(ethconfig.Defaults.Snapshot, snapshotLocation, 0, logger)
+
+	if err := activeSnapshots.ReopenFolder(); err != nil {
+		return HeimdallSimulator{}, err
+	}
+	if err := activeBorSnapshots.ReopenFolder(); err != nil {
+		return HeimdallSimulator{}, err
+	}
+
+	downloader, err := NewTorrentClient(ctx, chain, snapshotLocation, logger)
+	if err != nil {
+		return HeimdallSimulator{}, err
+	}
+
+	s := HeimdallSimulator{
+		ctx:                ctx,
+		knownSnapshots:     knownSnapshots,
+		activeSnapshots:    activeSnapshots,
+		knownBorSnapshots:  knownBorSnapshots,
+		activeBorSnapshots: activeBorSnapshots,
+		blockReader:        freezeblocks.NewBlockReader(activeSnapshots, activeBorSnapshots),
+		logger:             logger,
+		downloader:         downloader,
+		nextSpan:           0,
+	}
+
+	go func() {
+		<-ctx.Done()
+		s.Close()
+	}()
+
+	return s, nil
+}
+
+// FetchLatestSpan gets the next span from the snapshot
+func (h *HeimdallSimulator) FetchLatestSpan(ctx context.Context) (*heimdall.Span, error) {
+	span, err := h.getSpan(h.ctx, h.nextSpan)
```

Comment:
fetch latest span should return the last span in the latest (last) .seg file

Reply:
as mentioned in the following comment https://github.com/ledgerwatch/erigon/pull/9819/files#r1542056781 we can remove the nextSpan attribute and logic